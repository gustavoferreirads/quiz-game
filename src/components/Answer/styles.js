import { StyleSheet } from 'react-native';

import { colors, fonts, metrics } from '../../styles';

const borderWidth = 2;
const styles = StyleSheet.create({
  container: {
    width: '100%',
    borderRadius: 40,
    borderStyle: 'solid',
    borderWidth,
    borderColor: colors.lighter,
    alignItems: 'center',
    flexDirection: 'row',
    padding: metrics.size,
    marginTop: 20,
  },
  contentOption: {
    borderStyle: 'solid',
    borderRadius: 32,
    width: 32,
    height: 32,
    borderWidth,
    borderColor: colors.primary,
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: 10,
  },
  option: {
    fontSize: fonts.big,
    color: colors.primary,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  text: {
    fontSize: fonts.regular,
    fontWeight: 'bold',
    color: colors.regular,
    textTransform: 'uppercase',
  },
});

export default styles;
