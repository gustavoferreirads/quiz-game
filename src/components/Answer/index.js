import React from 'react';
import * as PropTypes from 'prop-types';
import { Text, TouchableOpacity, View } from 'react-native';
import styles from './styles';

const Answer = ({ option, text, onPress }) => (
  <TouchableOpacity style={styles.container} onPress={onPress}>
    <View style={styles.contentOption}>
      <Text style={styles.option}>{option}</Text>
    </View>
    <Text style={styles.text}>{text}</Text>
  </TouchableOpacity>
);

Answer.defaultProps = {};
Answer.propTypes = {
  option: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  onPress: PropTypes.func.isRequired,
};

export default Answer;
