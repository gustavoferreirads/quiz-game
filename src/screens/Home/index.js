import React from 'react';
import * as PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Text, View } from 'react-native';
import * as actions from '../../redux/actions/quiz';
import Container from '../../components/Container';
import Button from '../../components/Button';

import styles from './styles';

export const Home = ({ navigation, fetchQuestions }) => {
  const navigateBegin = () => {
    fetchQuestions();
    navigation.navigate('Quiz');
  };

  return (
    <Container style={styles.container}>
      <View style={styles.boxTitle}>
        <Text style={styles.title}>Welcome to the</Text>
        <Text style={styles.title}>Trivia Challenge!</Text>
      </View>
      <View style={styles.grow}>
        <Text style={styles.internalText}>
          You will be presented with 10 True or False questions.
        </Text>
      </View>
      <View style={styles.grow}>
        <Text style={[styles.internalText, styles.bolder]}>Can you score 100%?</Text>
      </View>
      <View style={styles.grow}>
        <Button onPress={navigateBegin}>
          <Text>BEGIN</Text>
        </Button>
      </View>
    </Container>
  );
};

Home.defaultProps = {
  navigation: {},
};

Home.propTypes = {
  navigation: PropTypes.object,
  fetchQuestions: PropTypes.func.isRequired,
};

export default connect(
  () => ({}),
  dispatch => bindActionCreators(actions, dispatch),
)(Home);
