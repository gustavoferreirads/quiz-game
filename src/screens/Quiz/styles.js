import { StyleSheet } from 'react-native';
import { colors, fonts, metrics } from '../../styles';

export default StyleSheet.create({
  container: {
    backgroundColor: colors.lighter,
    flex: 1,
    padding: metrics.mediumSize,
    justifyContent: 'flex-start',
  },
  title: {
    fontSize: fonts.largest,
    marginTop: metrics.margin,
    marginBottom: metrics.margin,
    color: colors.dark,
    textAlign: 'center',
  },
  content: {
    backgroundColor: colors.white,
    borderRadius: metrics.radius,
    height: 375,
    width: '100%',
  },
  box: {
    padding: metrics.size,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: colors.primary,
    height: 175,
    borderTopEndRadius: metrics.radius,
    borderTopLeftRadius: metrics.radius,
  },
  internalBox: {
    textAlign: 'center',
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
  },
  question: {
    color: colors.white,
    fontSize: fonts.big,
    fontWeight: 'bold',
    textTransform: 'uppercase',
    textAlign: 'center',
  },
  position: {
    color: colors.white,
    fontSize: fonts.big,
    fontWeight: 'bold',
  },
  contentAnswer: {
    padding: metrics.size,
    flex: 1,
  },
});
