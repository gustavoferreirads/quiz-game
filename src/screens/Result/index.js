import React, { Fragment } from 'react';
import * as PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { FlatList, Text, View } from 'react-native';
import ProgressCircle from 'react-native-progress-circle';
import { AntDesign } from '@expo/vector-icons/build/Icons';
import Button from '../../components/Button';
import Container from '../../components/Container';
import * as actions from '../../redux/actions/quiz';
import { icons, progressCircle, styles } from './styles';

const { radius, borderWidth, color, shadowColor, bgColor } = progressCircle;

export const Result = ({ navigation, fetchQuestions, answers, score, percent }) => {
  const navigateBegin = () => {
    navigation.navigate('Home');
  };

  const navigateQuiz = () => {
    fetchQuestions();
    navigation.navigate('Quiz');
  };

  return (
    <Container style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.title}>You scored</Text>

        <View style={styles.graph}>
          <ProgressCircle
            percent={percent}
            radius={radius}
            borderWidth={borderWidth}
            color={color}
            shadowColor={shadowColor}
            bgColor={bgColor}
          >
            <Text style={styles.title}>
              {score}/{answers.length}
            </Text>
          </ProgressCircle>
        </View>

        <Text style={styles.subTitle}>You got {percent}% of the answers correct</Text>
        <Text style={styles.subTitle}>Congratulations!</Text>
      </View>

      <View style={styles.content}>
        <FlatList
          data={answers}
          renderItem={({ item }) => (
            <View style={styles.item}>
              <View style={styles.headerItem}>
                <Text style={styles.description}>{item.question}</Text>
                <Text style={[styles.description, styles.answer]}>{item.correct_answer}</Text>
              </View>

              <View style={styles.result}>
                {item.isCorrect && (
                  <Fragment>
                    <Text style={styles.text}>You're right!</Text>
                    <AntDesign name={icons.like.name} size={icons.size} color={icons.like.color} />
                  </Fragment>
                )}

                {!item.isCorrect && (
                  <Fragment>
                    <Text style={styles.text}>You missed!</Text>
                    <AntDesign
                      name={icons.unlike.name}
                      size={icons.size}
                      color={icons.unlike.color}
                    />
                  </Fragment>
                )}
              </View>
            </View>
          )}
          keyExtractor={(item, index) => index.toString()}
        />
      </View>
      <View style={styles.actions}>
        <Button onPress={navigateQuiz}>
          <Text>PLAY AGAIN?</Text>
        </Button>

        <Button type="outline" onPress={navigateBegin}>
          <Text>HOME</Text>
        </Button>
      </View>
    </Container>
  );
};

Result.defaultProps = {
  answers: [],
  navigation: {},
  score: 0,
  percent: 0,
};

Result.propTypes = {
  navigation: PropTypes.object,
  answers: PropTypes.arrayOf(PropTypes.object),
  fetchQuestions: PropTypes.func.isRequired,
  score: PropTypes.number,
  percent: PropTypes.number,
};

export default connect(
  ({ Quiz }) => ({
    answers: Quiz.answers,
    score: Quiz.score,
    percent: Quiz.percent,
  }),
  dispatch => bindActionCreators(actions, dispatch),
)(Result);
