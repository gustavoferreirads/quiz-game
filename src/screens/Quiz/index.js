import React from 'react';
import * as PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Text, View } from 'react-native';
import Container from '../../components/Container';
import * as actions from '../../redux/actions/quiz';
import { withLoading } from '../../components/Loading';
import styles from './styles';
import Answer from '../../components/Answer';

const TRUE = 'True';
const FALSE = 'False';

export const Quiz = ({ navigation, current, size, doAnswer }) => {
  const { position, wrapQuestion } = current;

  const onAnswer = answer => {
    doAnswer(answer);
    if (position === size) {
      navigation.navigate('Results');
    }
  };

  return (
    <Container style={styles.container}>
      <Text style={styles.title}>{wrapQuestion.category}</Text>
      <View style={styles.content}>
        <View style={styles.box}>
          <View style={styles.internalBox}>
            <Text style={styles.question}>{wrapQuestion.question}</Text>
          </View>
          <Text style={styles.position}>
            {position} of {size}
          </Text>
        </View>
        <View style={styles.contentAnswer}>
          <Answer option="A" text="True" onPress={() => onAnswer(TRUE)} />
          <Answer option="B" text="False" onPress={() => onAnswer(FALSE)} />
        </View>
      </View>
    </Container>
  );
};

Quiz.defaultProps = {
  current: {},
  navigation: {},
  size: 0,
};

Quiz.propTypes = {
  navigation: PropTypes.object,
  doAnswer: PropTypes.func.isRequired,
  current: PropTypes.object,
  size: PropTypes.number,
};

export default withLoading(
  connect(
    store => ({
      current: store.Quiz.current,
      size: store.Quiz.size,
    }),
    dispatch => bindActionCreators(actions, dispatch),
  )(Quiz),
);
