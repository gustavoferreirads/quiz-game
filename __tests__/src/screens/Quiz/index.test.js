import React from 'react';
import { shallow } from 'enzyme';
import { Quiz } from '../../../../src/screens/Quiz';
import Answer from '../../../../src/components/Answer';
import { mock } from './mock';

let doAnswer;
let navigator;

const assert = (answer, position, size = 10) => {
  const wrapper = shallow(
    <Quiz navigation={navigator} current={mock} doAnswer={doAnswer} size={size} />,
  );
  wrapper
    .find(Answer)
    .at(position)
    .props()
    .onPress();
  expect(doAnswer.mock.calls.length).toBe(1);
  expect(doAnswer.mock.calls[0][0]).toBe(answer);
};

beforeEach(() => {
  doAnswer = jest.fn(() => {});
  navigator = {
    navigate: jest.fn(x => x),
  };
});

it('renders a Quiz ', () => {
  expect(
    shallow(<Quiz doAnswer={() => jest.fn} current={mock} navigation={{}} size={10} />),
  ).toMatchSnapshot();
});

it('simulate click on button "True"', () => {
  assert('True', 0);
});

it('simulate click on button "False" and go Result', () => {
  assert('False', 1, 1);
  expect(navigator.navigate.mock.calls[0][0]).toBe('Results');
  expect(navigator.navigate.mock.calls.length).toBe(1);
});
