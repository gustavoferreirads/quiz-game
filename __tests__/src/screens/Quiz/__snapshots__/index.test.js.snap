// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Quiz  1`] = `
<Container
  style={
    Object {
      "backgroundColor": "#eee",
      "flex": 1,
      "justifyContent": "flex-start",
      "padding": 20,
    }
  }
>
  <Component
    style={
      Object {
        "color": "#333",
        "fontSize": 22,
        "marginBottom": 40,
        "marginTop": 40,
        "textAlign": "center",
      }
    }
  >
    Entertainment: Film
  </Component>
  <Component
    style={
      Object {
        "backgroundColor": "#FFF",
        "borderRadius": 10,
        "height": 375,
        "width": "100%",
      }
    }
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FF9982",
          "borderTopEndRadius": 10,
          "borderTopLeftRadius": 10,
          "height": 175,
          "justifyContent": "center",
          "padding": 10,
        }
      }
    >
      <Component
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "textAlign": "center",
          }
        }
      >
        <Component
          style={
            Object {
              "color": "#FFF",
              "fontSize": 16,
              "fontWeight": "bold",
              "textAlign": "center",
              "textTransform": "uppercase",
            }
          }
        >
          &quot;Cube&quot;, &quot;Cube 2: Hypercube&quot; and &quot;Cube Zero&quot; were directed by the same person.
        </Component>
      </Component>
      <Component
        style={
          Object {
            "color": "#FFF",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        1
         of 
        10
      </Component>
    </Component>
    <Component
      style={
        Object {
          "flex": 1,
          "padding": 10,
        }
      }
    >
      <Answer
        onPress={[Function]}
        option="A"
        text="True"
      />
      <Answer
        onPress={[Function]}
        option="B"
        text="False"
      />
    </Component>
  </Component>
</Container>
`;
